nipoppy_new/nipoppy/workflows/base.py

```diff
@@ -0,0 +1,315 @@
+        ):
+            """Consume lines from an IO stream and append them to a string."""
+            for line in output_source:
+                line = line.strip("\n")
+                self.logger.log(level=log_level, msg=f"{log_prefix} {line}")
+            return output_str
+
+        # build command string
+        if not isinstance(command_or_args, str):
+            args = [str(arg) for arg in command_or_args]
+            command = shlex.join(args)
+        else:
+            command = command_or_args
+            args = shlex.split(command)
+
+        # only pass a single string if shell is True
+        if not kwargs.get("shell"):
+            command_or_args = args
+
+        self.log_command(command)
+
+        stdout_str = ""
+        stderr_str = ""
+        if not self.dry_run:
+            process = subprocess.Popen(
+                command_or_args,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                text=True,
+                **kwargs,
+            )
+
+            while process.poll() is None:
+                stdout_str = process_output(
+                    process.stdout,
+                    stdout_str,
+                    self.log_prefix_run_stdout,
+                )
+
+                stderr_str = process_output(
+                    process.stderr,
+                    stderr_str,
+                    self.log_prefix_run_stderr,
+                    log_level=logging.ERROR,
+                )
+
+            if check and process.returncode != 0:
+                exception = subprocess.CalledProcessError(process.returncode, command)
+                raise exception
+
+            run_output = process
+
+        else:
+            run_output = command
+
+        return run_output
+
+    def save_tabular_file(self, tabular: BaseTabular, fpath: Path):
+        """Save a tabular file."""
+        if not self.dry_run:
+            fpath_doughnut_backup = tabular.save_with_backup(fpath)
+        if fpath_doughnut_backup is not None:
+            self.logger.info(f"Saved to {fpath} (-> {fpath_doughnut_backup})")
+        else:
+            self.logger.info(f"No changes to file at {fpath}")
+
+    def run_setup(self, **kwargs):
+        """Run the setup part of the workflow."""
+        self.logger.info(f"========== BEGIN {self.name.upper()} WORKFLOW ==========")
+        self.logger.info(self)
+        if self.dry_run:
+            self.logger.info("Doing a dry run")
+        if self.validate_layout:
+            try:
+                self.layout.validate()
+            except FileNotFoundError as exception:
+                raise RuntimeError(
+                    f"Dataset does not follow expected directory structure: {exception}"
+                )
+
+    @abstractmethod
+    def run_main(self, **kwargs):
+        """Run the main part of the workflow."""
+        pass
+
+    def run_cleanup(self, **kwargs):
```

Comment:
Right now they are overridden in child classes. I didn't make this or `run_setup` abstract because 1) they are not mandatory and 2) there is a default behaviour